src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -334,7 +336,8 @@ private List<FlightState> getAccessibleFlights(
       }
       filterResults.forEach(
           flightState -> {
-            if (userLaunchedFlight(flightState, userReq)) {
+            if ((canListAnyJob || userLaunchedFlight(flightState, userReq))
+                && jobIdSet.contains(flightState.getFlightId())) {
```

Comment:
I think that in the current state, if `jobIdSet` is empty, then we do not want to filter down our results by job ID, since no flight ID will be present in an empty set.

For this reason, I don't agree with making the controller pass an empty collection when the jobIds input is undefined, since the service's behavior on undefined input will be different.  Maybe instead of passing a null through, an `Optional` would be a better fit.  What do you think?

Reply:
You're totally right, I was thinking the id set was an exclusion list, not an inclusion list. Maybe the easiest fix is to make a lambda for the test outside the `while` and use it in both `if's. It looks like the purpose of this check is to guard additions to `flightStateList`, so something like this could work:

```
    List<FlightState> flightStateList = new ArrayList<>();
    Consumer<FlightState> addToList =
        flightState -> {
          if (jobIdSet.isEmpty() || jobIdSet.contains(flightState.getFlightId())) {
            flightStateList.add(flightState);
          }
        };
    while (flightStateList.size() < offset + limit) {
        ...
            if ((canListAnyJob || userLaunchedFlight(flightState, userReq))) {
              addToList.accept(flightState);
            } else {
        ...
                if (userRoles.contains(action.toString())) {
                  addToList.accept(flightState);
                }
        ...
```
I think this is nice in that it removes the common code completely. Another approach is to compute the `boolean` value inside the `while` and use the same value in both `if's.

Reply:
I actually like the optional approach for this case.  It's a bit more explicit in the signature (e.g. it indicates that it's expected that it might be empty and that there's special handling for that). And the code would like something like:
```suggestion
                && jobIdSet.map(j -> j.contains(flightState.getFlightId())).orElse(true)) {
```

which is nice and concise

Reply:
Though hold on, we've got precedent.  We should be consistent.  We do this in enumerateSnapshot (we accept a list of datasetIds) and in that case, we have the pattern that if a list is empty then we assume that no filtering is needed (which seems reasonable).  In that case, it would look more like:
```suggestion
                && (jobIdSet.isEmpty() || jobIdSet.contains(flightState.getFlightId()))) {
```

Reply:
Either of these approaches could be used with the `Consumer` lambda in my suggestion to avoid code duplication.

Or here's an even more radical approach; override `add` in `ArrayList`:
```
    List<FlightState> flightStateList = new ArrayList<>() {
      public boolean add(FlightState flightState) {
        if (jobIdSet.isEmpty() || jobIdSet.contains(flightState.getFlightId())) {
          return super.add(flightState);
        }
        return false;
      }
    };
    while (flightStateList.size() < offset + limit) {
        ...
            if ((canListAnyJob || userLaunchedFlight(flightState, userReq))) {
              flightStateList.add(flightState);
            } else {
        ...
                if (userRoles.contains(action.toString())) {
                  flightStateList.add(flightState);
                }
        ...
```

Reply:
hmm...I'm not seeing the duplication...

Reply:
My main concern with the consumer approach is that it make it harder I find to inspect the code and I guess I'm not seeing the duplication

Reply:
The duplication is `if (jobIdSet.isEmpty() || jobIdSet.contains(flightState.getFlightId()) { flightStateList.add(flightState); }` which is the same in both places. If the logic is changed it will need to be changed in both places. But it's not _that_ much duplication so I'd go with what you find easiest to read.